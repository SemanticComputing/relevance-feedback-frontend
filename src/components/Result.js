import React from 'react';
import { object, func } from 'prop-types';
import { Row, Col, Button } from 'reactstrap';
import FontAwesome from 'react-fontawesome';
import Topics from '../components/Topics';
import './Result.css';

const Result = ({ result, updateThumb, describeTopic }) => {

  const handleThumbsUp = () => updateThumb({ result, value: true });
  const handleThumbsDown = () => updateThumb({ result, value: false });

  let thumbsUpStyle = {};
  let thumbsDownStyle = {};
  if (result.thumb === true) {
    thumbsUpStyle.color = 'green';
  } else if (result.thumb === false) {
    thumbsDownStyle.color = 'red';
  }

  return (
    <Row className="result">
      <Col className="col-sm-1">
        <Topics topics={result.topic} describeTopic={describeTopic} />
      </Col>
      <Col className="col-sm-8">
        <Row>
          <Col><a href={result.url}>{result.name}</a></Col>
        </Row>
        <Row>
          <Col>{result.description}</Col>
        </Row>
      </Col>
      <Col className="col-sm-3">
        <Row>
          <Col><Button onClick={handleThumbsUp}><FontAwesome style={thumbsUpStyle} className="thumbs thumbs-up" size="2x" name="thumbs-up" /></Button></Col>
          <Col><Button onClick={handleThumbsDown}><FontAwesome style={thumbsDownStyle} className="thumbs thumbs-down" size="2x" name="thumbs-down" /></Button></Col>
        </Row>
      </Col>
    </Row>
  );
};

Result.propTypes = {
  result: object,
  updateThumb: func,
  describeTopic: func
};

export default Result;
