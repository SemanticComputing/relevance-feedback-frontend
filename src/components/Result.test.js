import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import Result from './Result';

const updateThumb = jest.fn();
const describeTopic = jest.fn();
const result = {
  name: 'title',
  url: 'http://example.fi',
  description: 'description'
};

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<Result currentTopic={{}} describeTopic={describeTopic} updateThumb={updateThumb} result={result} />, div);
});

it('renders a result', () => {
  const tree = renderer
    .create(<Result currentTopic={{}} describeTopic={describeTopic} updateThumb={updateThumb} result={result} />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('indicates a thumb up', () => {
  const thumbResult = Object.assign({}, result, { thumb: true });
  const tree = renderer
    .create(<Result currentTopic={{}} describeTopic={describeTopic} updateThumb={updateThumb} result={thumbResult} />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('indicates a thumb down', () => {
  const thumbResult = Object.assign({}, result, { thumb: false });
  const tree = renderer
    .create(<Result currentTopic={{}} describeTopic={describeTopic} updateThumb={updateThumb} result={thumbResult} />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('calls updateThumb when a thumb is clicked', () => {
  const resultElem = mount(<Result currentTopic={{}} describeTopic={describeTopic} updateThumb={updateThumb} result={result} />);
  resultElem.find({ name: 'thumbs-up' }).simulate('click');

  expect(updateThumb.mock.calls.length).toBe(1);
  expect(updateThumb.mock.calls[0][0]).toEqual({ result, value: true });

  resultElem.find({ name: 'thumbs-down' }).simulate('click');

  expect(updateThumb.mock.calls.length).toBe(2);
  expect(updateThumb.mock.calls[1][0]).toEqual({ result, value: false });
});
