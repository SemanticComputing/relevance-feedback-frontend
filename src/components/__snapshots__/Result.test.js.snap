// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`indicates a thumb down 1`] = `
<div
  className="row"
>
  <div
    className="col-sm-8 col"
  >
    <div
      className="row"
    >
      <div
        className="col"
      >
        <a
          href="http://example.fi"
        >
          title
        </a>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col"
      >
        description
      </div>
    </div>
  </div>
  <div
    className="col-sm-2 col"
  >
    <div
      className="row"
    >
      <div
        className="col"
      >
        <button
          className="btn btn-secondary"
          onClick={[Function]}
          type="button"
        >
          <span
            aria-hidden={true}
            className="fa fa-thumbs-up fa-2x thumbs thumbs-up"
            style={Object {}}
          />
        </button>
      </div>
      <div
        className="col"
      >
        <button
          className="btn btn-secondary"
          onClick={[Function]}
          type="button"
        >
          <span
            aria-hidden={true}
            className="fa fa-thumbs-down fa-2x thumbs thumbs-down"
            style={
              Object {
                "color": "red",
              }
            }
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`indicates a thumb up 1`] = `
<div
  className="row"
>
  <div
    className="col-sm-8 col"
  >
    <div
      className="row"
    >
      <div
        className="col"
      >
        <a
          href="http://example.fi"
        >
          title
        </a>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col"
      >
        description
      </div>
    </div>
  </div>
  <div
    className="col-sm-2 col"
  >
    <div
      className="row"
    >
      <div
        className="col"
      >
        <button
          className="btn btn-secondary"
          onClick={[Function]}
          type="button"
        >
          <span
            aria-hidden={true}
            className="fa fa-thumbs-up fa-2x thumbs thumbs-up"
            style={
              Object {
                "color": "green",
              }
            }
          />
        </button>
      </div>
      <div
        className="col"
      >
        <button
          className="btn btn-secondary"
          onClick={[Function]}
          type="button"
        >
          <span
            aria-hidden={true}
            className="fa fa-thumbs-down fa-2x thumbs thumbs-down"
            style={Object {}}
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders a result 1`] = `
<div
  className="row"
>
  <div
    className="col-sm-8 col"
  >
    <div
      className="row"
    >
      <div
        className="col"
      >
        <a
          href="http://example.fi"
        >
          title
        </a>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col"
      >
        description
      </div>
    </div>
  </div>
  <div
    className="col-sm-2 col"
  >
    <div
      className="row"
    >
      <div
        className="col"
      >
        <button
          className="btn btn-secondary"
          onClick={[Function]}
          type="button"
        >
          <span
            aria-hidden={true}
            className="fa fa-thumbs-up fa-2x thumbs thumbs-up"
            style={Object {}}
          />
        </button>
      </div>
      <div
        className="col"
      >
        <button
          className="btn btn-secondary"
          onClick={[Function]}
          type="button"
        >
          <span
            aria-hidden={true}
            className="fa fa-thumbs-down fa-2x thumbs thumbs-down"
            style={Object {}}
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
